src/input.rs

```diff
@@ -135,233 +128,257 @@ impl State {
             InputEvent::Keyboard { event, .. } => {
                 let serial = SERIAL_COUNTER.next_serial();
                 let time = Event::time_msec(&event);
+                let pressed = event.state() == KeyState::Pressed;
 
-                let mut action = self.niri.seat.get_keyboard().unwrap().input(
+                let Some(Some(action)) = self.niri.seat.get_keyboard().unwrap().input(
                     self,
                     event.key_code(),
                     event.state(),
                     serial,
                     time,
-                    |self_, mods, keysym| {
-                        if event.state() == KeyState::Pressed {
-                            let config = self_.niri.config.borrow();
-                            action(&config, comp_mod, keysym, *mods).into()
-                        } else {
-                            FilterResult::Forward
+                    |this, mods, keysym| {
+                        let config = this.niri.config.borrow();
+                        let key_code = event.key_code();
+                        match (
+                            action(&config, comp_mod, keysym, *mods),
+                            this.niri.suppress_keys.entry(key_code),
+                        ) {
+                            (action, Entry::Occupied(mut entry)) => {
+                                let key = entry.get_mut();
+                                if pressed {
+                                    *key += 1;
+                                } else {
+                                    *key -= 1;
+                                }
```

Comment:
>Can we legitimately go above 1 here? I'm thinking it might be better to use a simple bool instead of a counter.

yes, with keyboard having duplicated keys.

>Also, for the VT switch keys like Ctrl-Alt-F1, I'm not sure we ever see key release events. A counter will keep counting up unnecessarily, while a bool will just stay true.

I think the only solution for that is to `clear` the `HashMap`.